// osu-api-v1-js/dist/beatmap.js
/**
 * For the `approved` of a `Beatmap` (for example, `Categories[beatmap.approved]` would return "RANKED" if 1) https://osu.ppy.sh/wiki/en/Beatmap/Category
 */
export var Categories;
(function (Categories) {
    Categories[Categories["GRAVEYARD"] = -2] = "GRAVEYARD";
    Categories[Categories["WIP"] = -1] = "WIP";
    Categories[Categories["PENDING"] = 0] = "PENDING";
    Categories[Categories["RANKED"] = 1] = "RANKED";
    Categories[Categories["APPROVED"] = 2] = "APPROVED";
    Categories[Categories["QUALIFIED"] = 3] = "QUALIFIED";
})(Categories || (Categories = {}));
/**
 * For the `genre_id` of a `Beatmap` (for example, `Genres[beatmap.genre_id]` would return "NOVELTY" if 7)
 */
export var Genres;
(function (Genres) {
    Genres[Genres["ANY"] = 0] = "ANY";
    Genres[Genres["UNSPECIFIED"] = 1] = "UNSPECIFIED";
    Genres[Genres["VIDEO GAME"] = 2] = "VIDEO GAME";
    Genres[Genres["ANIME"] = 3] = "ANIME";
    Genres[Genres["ROCK"] = 4] = "ROCK";
    Genres[Genres["POP"] = 5] = "POP";
    Genres[Genres["OTHER"] = 6] = "OTHER";
    Genres[Genres["NOVELTY"] = 7] = "NOVELTY";
    Genres[Genres[""] = 8] = "";
    Genres[Genres["HIP HOP"] = 9] = "HIP HOP";
    Genres[Genres["ELECTRONIC"] = 10] = "ELECTRONIC";
    Genres[Genres["METAL"] = 11] = "METAL";
    Genres[Genres["CLASSICAL"] = 12] = "CLASSICAL";
    Genres[Genres["FOLK"] = 13] = "FOLK";
    Genres[Genres["JAZZ"] = 14] = "JAZZ";
})(Genres || (Genres = {}));
/**
 * For the `language_id` of a `Beatmap` (for example, `Languages[beatmap.language_id]` would return "FRENCH" if 7)
 */
export var Languages;
(function (Languages) {
    Languages[Languages["ANY"] = 0] = "ANY";
    Languages[Languages["UNSPECIFIED"] = 1] = "UNSPECIFIED";
    Languages[Languages["ENGLISH"] = 2] = "ENGLISH";
    Languages[Languages["JAPANESE"] = 3] = "JAPANESE";
    Languages[Languages["CHINESE"] = 4] = "CHINESE";
    Languages[Languages["INSTRUMENTAL"] = 5] = "INSTRUMENTAL";
    Languages[Languages["KOREAN"] = 6] = "KOREAN";
    Languages[Languages["FRENCH"] = 7] = "FRENCH";
    Languages[Languages["GERMAN"] = 8] = "GERMAN";
    Languages[Languages["SWEDISH"] = 9] = "SWEDISH";
    Languages[Languages["SPANISH"] = 10] = "SPANISH";
    Languages[Languages["ITALIAN"] = 11] = "ITALIAN";
    Languages[Languages["RUSSIAN"] = 12] = "RUSSIAN";
    Languages[Languages["POLISH"] = 13] = "POLISH";
    Languages[Languages["OTHER"] = 14] = "OTHER";
})(Languages || (Languages = {}));
